### PR TITLE
bpo-38031: Fix a possible assertion failure in _io.FileIO()

### PR DESCRIPTION
**Pull request by ZackerySpytz**
Use PyErr_Fetch() / _PyErr_ChainExceptions() when calling
internal_close() on error.  The opener may return an invalid
fd, which will cause the close() call in internal_close() to fail.

https://bugs.python.org/issue38031

### CODE REVIEW

**Review by aeros — COMMENTED** [review]
Thanks for the PR @ZackerySpytz.

The macOS failures in Azure may be related to https://bugs.python.org/issue37245 and not related to this PR. (Build log: https://dev.azure.com/Python/cpython/_build/results?buildId=49868&view=logs&j=18d1a34d-6940-5fc1-f55b-405e2fba32b1)

``
2 tests failed:
    test_functools test_multiprocessing_spawn
``

Both of the above tests timed out.

File: Lib/test/test_io.py [hunk 1]
```diff
@@ -875,6 +875,13 @@ def badopener(fname, flags):
             open('non-existent', 'r', opener=badopener)
         self.assertEqual(str(cm.exception), 'opener returned -2')
 
+    def test_opener_invalid_fd(self):
+        self.open(support.TESTFN, 'w').close()
+        fd = os.open(support.TESTFN, os.O_RDONLY)
+        os.close(fd)
+        with self.assertRaisesRegex(OSError, 'Bad file descriptor'):
```
Comment (vstinner) [inline Lib/test/test_io.py]: I would prefer to check exc.errno == errno.EBADF, I'm not sure that all platforms use exactly the same spelling for os.strerror(errno.EBADF).

File: Lib/test/test_io.py [hunk 2]
```diff
@@ -875,6 +875,13 @@ def badopener(fname, flags):
             open('non-existent', 'r', opener=badopener)
         self.assertEqual(str(cm.exception), 'opener returned -2')
 
+    def test_opener_invalid_fd(self):
+        self.open(support.TESTFN, 'w').close()
```
Comment (vstinner) [inline Lib/test/test_io.py]: You may use __file__ rather than creating TESTFN.

File: Lib/test/test_io.py [hunk 3]
```diff
@@ -875,6 +875,13 @@ def badopener(fname, flags):
...
+        fd = os.open(support.TESTFN, os.O_RDONLY)
+        os.close(fd)
```
Comment (vstinner) [inline Lib/test/test_io.py]: Maybe add a comment like:

```
# fd is now an invalid file descriptor
```

**Comment by vstinner** [comment]
cc @serhiy-storchaka @pitrou

File: Lib/test/test_io.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/test/test_io.py]: `__file__` does not work in exotic cases: for example when the testsuite is imported from a ZIP archive.

**Review by serhiy-storchaka — COMMENTED** [review]
Mostly LGTH. Just address Victor's comment.

File: Lib/test/test_io.py (hunk 2 above)
Comment (aeros) [reply Lib/test/test_io.py]: @serhiy-storchaka:
> ``__file__`` does not work in exotic cases: for example when the testsuite is imported from a ZIP archive.

Is this noteworthy enough to briefly mention it in the [documentation for ``__file__``](https://docs.python.org/3/reference/import.html#__file__)? I see no mention of it there. Also, are there any other edge cases where ``__file__`` is not usable that you can think of?

Comment (serhiy-storchaka) [reply Lib/test/test_io.py]: It is obviously: if the sources were loaded not from a file, `__file__` does not refer to a real file. Other example: builtin modules (like `sys`). Third-party loaders can provide other examples.

Comment (aeros) [reply Lib/test/test_io.py]: > It is obviously: if the sources were loaded not from a file, __file__ does not refer to a real file. Other example: builtin modules (like sys). Third-party loaders can provide other examples.

Thanks for the clarification. I was aware of the first example, but not the limitation with builtin modules.

**Comment by csabella** [comment]
@ZackerySpytz, please address the code review.  Thanks!

File: Lib/test/test_io.py (hunk 3 above)
Comment (iritkatriel) [inline Lib/test/test_io.py]:
```suggestion
        os.close(fd)
        # fd is now an invalid file descriptor
```

File: Lib/test/test_io.py (hunk 1 above)
Comment (iritkatriel) [inline Lib/test/test_io.py]:
```suggestion
        with self.assertRaises(OSError):
            self.assertEqual(exc.errno, errno.EBADF)
```

**Review by iritkatriel — COMMENTED** [review]
Converted @vstinner's comment into suggestions.

File: Lib/test/test_io.py [hunk 4]
```diff
@@ -888,6 +888,15 @@ def badopener(fname, flags):
...
+        # fd is now an invalid file descriptor
+        with self.assertRaises(OSError):
+            self.assertEqual(exc.errno, errno.EBADF)
```
Comment (iritkatriel) [inline Lib/test/test_io.py]:
```suggestion
        with self.assertRaises(OSError) as cm:
            self.assertEqual(cm.exception.errno, errno.EBADF)
```

File: Lib/test/test_io.py [hunk 5]
```diff
@@ -888,6 +888,15 @@ def badopener(fname, flags):
             open('non-existent', 'r', opener=badopener)
         self.assertEqual(str(cm.exception), 'opener returned -2')
 
+    def test_opener_invalid_fd(self):
+        self.open(support.TESTFN, 'w').close()
```
Comment (iritkatriel) [inline Lib/test/test_io.py]:
```suggestion
        self.open(os_helper.TESTFN, 'w').close()
```

File: Lib/test/test_io.py [hunk 6]
```diff
@@ -888,6 +888,15 @@ def badopener(fname, flags):
...
+        fd = os.open(support.TESTFN, os.O_RDONLY)
```
Comment (iritkatriel) [inline Lib/test/test_io.py]:
```suggestion
        fd = os.open(os_helper.TESTFN, os.O_RDONLY)
```

File: Lib/test/test_io.py [hunk 7]
```diff
@@ -888,6 +888,15 @@ def badopener(fname, flags):
...
         self.assertEqual(str(cm.exception), 'opener returned -2')
 
+    def test_opener_invalid_fd(self):
+        self.open(os_helper.TESTFN, 'w').close()
+        fd = os.open(os_helper.TESTFN, os.O_RDONLY)
+        os.close(fd)
+        # fd is now an invalid file descriptor
+        with self.assertRaises(OSError) as cm:
+            self.assertEqual(cm.exception.errno, errno.EBADF)
+            self.open('foo', opener=lambda name, flags: fd)
```
Comment (iritkatriel) [inline Lib/test/test_io.py]:
```suggestion
        with self.assertRaises(OSError) as cm:
            self.open('foo', opener=lambda name, flags: fd)
        self.assertEqual(cm.exception.errno, errno.EBADF)
```

File: Lib/test/test_io.py [hunk 8]
```diff
@@ -888,6 +888,15 @@ def badopener(fname, flags):
             open('non-existent', 'r', opener=badopener)
         self.assertEqual(str(cm.exception), 'opener returned -2')
 
+    def test_opener_invalid_fd(self):
+        self.open(os_helper.TESTFN, 'w').close()
+        fd = os.open(os_helper.TESTFN, os.O_RDONLY)
+        os.close(fd)
```
Comment (vstinner) [inline Lib/test/test_io.py]: You can use os_helper.make_bad_fd(), from ``from test.support import os_helper``.

File: Lib/test/test_io.py [hunk 9]
```diff
@@ -888,6 +888,15 @@ def badopener(fname, flags):
             open('non-existent', 'r', opener=badopener)
         self.assertEqual(str(cm.exception), 'opener returned -2')
 
+    def test_opener_invalid_fd(self):
```
Comment (vstinner) [inline Lib/test/test_io.py:891]: Would you mind to add a reference to the issue (gh-15688) and explain why this test was added? Like:

```
gh-15688: Check that OSError is raised with EBADF error code if the opener returns an invalid file descriptor.
```

File: Lib/test/test_io.py [hunk 10]
```diff
@@ -888,6 +888,15 @@ def badopener(fname, flags):
...
+        self.open(os_helper.TESTFN, 'w').close()
+        fd = os.open(os_helper.TESTFN, os.O_RDONLY)
+        os.close(fd)
+        # fd is now an invalid file descriptor
+        with self.assertRaises(OSError) as cm:
```
Comment (iritkatriel) [inline Lib/test/test_io.py]:
```suggestion
    def test_opener_invalid_fd(self):
        # Check that OSError is raised with EBADF error code if the opener returns an invalid
        # file descriptor (see gh-15688).
        
        fd = os_helper.make_bad_fd()
        with self.assertRaises(OSError) as cm:
```

File: Lib/test/test_io.py [hunk 11]
```diff
@@ -888,6 +888,15 @@ def badopener(fname, flags):
             open('non-existent', 'r', opener=badopener)
         self.assertEqual(str(cm.exception), 'opener returned -2')
 
+    def test_opener_invalid_fd(self):
+        # Check that OSError is raised with EBADF error code if the opener returns an invalid
+        # file descriptor (see gh-15688).
```
Comment (iritkatriel) [inline Lib/test/test_io.py]:
```suggestion
        # file descriptor (see gh-82212).
```

**Review by vstinner — APPROVED** [review]
LGTM. Thanks for updating the PR @iritkatriel ;-)

File: Lib/test/test_io.py [hunk 12]
```diff
@@ -888,6 +888,15 @@ def badopener(fname, flags):
             open('non-existent', 'r', opener=badopener)
         self.assertEqual(str(cm.exception), 'opener returned -2')
 
+    def test_opener_invalid_fd(self):
+        # Check that OSError is raised with EBADF error code if the opener returns an invalid
+        # file descriptor (see gh-82212).
+
```
Comment (iritkatriel) [inline Lib/test/test_io.py]:
```suggestion

```

**Pull request merged**